File: src/Functional/Clojure/prototype.clj
(defn clone-profile [profile]
  (assoc profile :features (vec (:features profile))))

(defn describe [profile]
  (str (:name profile) ": " (clojure.string/join "," (:features profile))))

(let [original {:name "orders" :features ["metrics"]}
      base-clone (clone-profile original)
      canary (-> base-clone
                 (assoc :name "orders-canary")
                 (update :features conj "tracing"))]
  (println (str "original=" (describe original)))
  (println (str "clone=" (describe canary))))

File: src/Functional/Haskell/Prototype.hs
module Main where

import Data.List (intercalate)

data ServiceProfile = ServiceProfile
  { profileName :: String,
    profileFeatures :: [String]
  }

cloneProfile :: ServiceProfile -> ServiceProfile
cloneProfile profile = profile {profileFeatures = map id (profileFeatures profile)}

describe :: ServiceProfile -> String
describe profile = profileName profile ++ ": " ++ intercalate "," (profileFeatures profile)

main :: IO ()
main = do
  let original = ServiceProfile "orders" ["metrics"]
      baseClone = cloneProfile original
      canary = baseClone {profileName = "orders-canary", profileFeatures = profileFeatures baseClone ++ ["tracing"]}
  putStrLn $ "original=" ++ describe original
  putStrLn $ "clone=" ++ describe canary
